### ISSUE TITLE
_start_response() got an unexpected keyword argument 'exc_info' on Heroku

### ISSUE THREAD

**Issue by daneoshiga** [encode/apistar#504]
I'm getting a strange exception in production and can't trace to what is going on.
can't reproduce locally, happening only on heroku, any idea of how to find it's cause?

```
│2018-05-02T16:20:05.126068+00:00 app[web.1]: [2018-05-02 16:20:05 +0000] [11] [ERROR] Error handling request /hooks/                                                                 │
│2018-05-02T16:20:05.126071+00:00 app[web.1]: Traceback (most recent call last):                                                                                                               │
│2018-05-02T16:20:05.126074+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/app.py", line 215, in __call__                                            │
│2018-05-02T16:20:05.126075+00:00 app[web.1]:     return self.injector.run(funcs, state)                                                                                                       │
│2018-05-02T16:20:05.126077+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/injector.py", line 106, in run                                            │
│2018-05-02T16:20:05.126079+00:00 app[web.1]:     state[output_name] = func(**func_kwargs)                                                                                                     │
│2018-05-02T16:20:05.126080+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/app.py", line 206, in __call__                                            │
│2018-05-02T16:20:05.126082+00:00 app[web.1]:     return self.injector.run(funcs, state)                                                                                                       │
│2018-05-02T16:20:05.126083+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/injector.py", line 106, in run                                            │
│2018-05-02T16:20:05.126085+00:00 app[web.1]:     state[output_name] = func(**func_kwargs)                                                                                                     │
│2018-05-02T16:20:05.126086+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/app.py", line 171, in finalize_wsgi                                       │
│2018-05-02T16:20:05.126088+00:00 app[web.1]:     exc_info=response.exc_info                                                                                                                   │
│2018-05-02T16:20:05.126090+00:00 app[web.1]: TypeError: _start_response() got an unexpected keyword argument 'exc_info'                                                                       │
│2018-05-02T16:20:05.126092+00:00 app[web.1]:                                                                                                                                                  │
│2018-05-02T16:20:05.126094+00:00 app[web.1]: During handling of the above exception, another exception occurred:                                                                              │
│2018-05-02T16:20:05.126095+00:00 app[web.1]:                                                                                                                                                  │
│2018-05-02T16:20:05.126096+00:00 app[web.1]: Traceback (most recent call last):                                                                                                               │
│2018-05-02T16:20:05.126098+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 135, in handle                                           │
│2018-05-02T16:20:05.126100+00:00 app[web.1]:     self.handle_request(listener, req, client, addr)                                                                                             │
│2018-05-02T16:20:05.126102+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 176, in handle_request                                   │
│2018-05-02T16:20:05.126103+00:00 app[web.1]:     respiter = self.wsgi(environ, resp.start_response)                                                                                           │
│2018-05-02T16:20:05.126105+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/newrelic/api/web_transaction.py", line 1282, in _nr_wsgi_application_wrapper_            │
│2018-05-02T16:20:05.126107+00:00 app[web.1]:     result = wrapped(environ, _start_response)                                                                                                   │
│2018-05-02T16:20:05.126109+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/app.py", line 222, in __call__                                            │
│2018-05-02T16:20:05.126110+00:00 app[web.1]:     return self.injector.run(funcs, state)                                                                                                       │
│2018-05-02T16:20:05.126112+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/injector.py", line 106, in run                                            │
│2018-05-02T16:20:05.126114+00:00 app[web.1]:     state[output_name] = func(**func_kwargs)                                                                                                     │
│2018-05-02T16:20:05.126115+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/apistar/server/app.py", line 171, in finalize_wsgi                                       │
│2018-05-02T16:20:05.126116+00:00 app[web.1]:     exc_info=response.exc_info                                                                                                                   │
│2018-05-02T16:20:05.127241+00:00 app[web.1]: TypeError: _start_response() got an unexpected keyword argument 'exc_info'                                                                       │
│2018-05-02T16:20:05.127751+00:00 app[web.1]: 10.79.192.24 - - [02/May/2018:16:20:05 +0000] "POST /hooks/ HTTP/1.1" 500 - "-" "-"                                                     │
```

**Comment by tino** [comment]
I can reproduce this locally with `uwsg --module app:app --http-socket :8000`

Commenting out https://github.com/encode/apistar/blob/master/apistar/server/app.py#L171 makes it work again

**Comment by daneoshiga** [comment]
I'm using gunicorn..

**Comment by rhelms** [comment]
At a guess, the start_response handler isn't accepting the exc_info parameter.

Both Official WSGI Specifications (referenced here: https://www.fullstackpython.com/wsgi-servers.html) specify that exc_info is an optional keyword argument, but confusingly, it mentions `As with all WSGI callables, the arguments must be supplied positionally, not by keyword.`, and then goes on to show examples of exc_info being used as a keyword argument.

However, looking at GUnicorm, it clearly accepts that parameter (https://github.com/benoitc/gunicorn/blob/e73ca252f7e1d0286998a0ae4254164291020a0c/gunicorn/http/wsgi.py#L233).

I wonder if Heroku is actually passing off to GUnicorn, or if something else is intercepting the WSGIStartResponse and replacing it with something that doesn't honor the expected interface of start_response.

**Comment by rhelms** [comment]
And here's the issue:
https://github.com/edmorley/newrelic-python-agent/blob/b5be319dcdf5dc054704a364fe56fef4774c5e8e/newrelic/newrelic/api/web_transaction.py#L1244

newrelic doesn't appear to honor the start_response interface. Perhaps it's just a typo that they have *args instead of **kwargs.

**Comment by daneoshiga** [comment]
I'm gonna fill a bug-report in new relic support center linking to this issue

**Comment by daneoshiga** [comment]
new relic support response

> Hi Danilo-
> 
> Thanks for brining this issue to our attention.
> 
> I discussed what you provided with our engineering team. We feel this is a bug in your web framework.
> 
> The WSGI specification says...
> 
> The second parameter passed to the application object is a callable of the form start_response(status, response_headers, exc_info=None). (As with all WSGI callables, the arguments must be supplied positionally, not by keyword.)
> 
> https://www.python.org/dev/peps/pep-3333/#the-start-response-callable
> 
> It appears that the framework is attempting to call start_response using keyword arguments here -
> 
> https://github.com/encode/apistar/blob/bd79eac59f535a80ffbe7d2ae0179c4262ccaa7a/apistar/server/app.py#L171 
> The issue you're seeing seems to be the result of what the framework is doing and not related to the Python agent.
> 
> Please let us know if you have any questions.
> 
> Regards,

**Comment by rhelms** [comment]
Maybe it is a bug in apistar. Perhaps the notation used is indicating an optional positional argument, rather than a keyword argument.

**Comment by tomchristie** [comment]
Thanks! Confirmed this in https://www.python.org/dev/peps/pep-0333/#the-start-response-callable...

> The second parameter passed to the application object is a callable of the form start_response(status, response_headers, exc_info=None). (As with all WSGI callables, the arguments must be supplied positionally, not by keyword.)

**Comment by tomchristie** [comment]
Closed via #506.
Released as 0.5.12.
